File: Python/32-LongestValidParentheses.py
class Solution:
    # @param {string} s
    # @return {integer}
    def longestValidParentheses(self, s):
        length = len(s)
        stack  = []
        maxlen = 0; last = -1
        for i in xrange(length):
            if s[i] == '(':
               stack.append(i)
            else:
               if stack == []:
                  last = i
               else:
                  stack.pop()
                  if stack == []:
                     maxlen = max(maxlen, i - last)
                  else:
                     maxlen = max(maxlen, i - stack[-1])
        return maxlen

if __name__ == '__main__':
   s = Solution()
   print s.longestValidParentheses(")()())")
   print s.longestValidParentheses("(()")
   print s.longestValidParentheses("()")

        

File: C++/32-LongestValidParentheses.cpp
#include "include/common.h"

// Chris : NTR

class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int> stack;
        int last = -1, maxLen = 0, end = 0;
        for (int i = 0; i < s.length(); i++) {
             if (s[i] == '(') {
                 stack.push_back(i);
             }
             else {
                 if (stack.empty()) {
                     last = i;
                 }
                 else {
                     stack.pop_back();
                     if (stack.empty()) {
                         maxLen = maxLen > i - last ? maxLen : i - last;
                     }
                     else {
                         // end = stack[stack.size() - 1];
                         maxLen = maxLen > i - stack[stack.size() - 1] ? maxLen : i - stack[stack.size() - 1];
                     };
                 };
             }
        };
        return maxLen;
    }
};

int main(void) {
    Solution s;
    cout << s.longestValidParentheses("()()") << endl;
    cout << s.longestValidParentheses("(()") << endl;
    cout << s.longestValidParentheses(")()()(") << endl;
    return 0;
};